wip: Optimize get_all_models

torisetxd opened the pull request:
# Pull Request Checklist

### Note to first-time contributors: Please open a discussion post in [Discussions](https://github.com/open-webui/open-webui/discussions) and describe your changes before submitting a pull request.

**Before submitting, make sure you've checked the following:**

- [x] **Target branch:** Please verify that the pull request targets the `dev` branch.
- [x] **Description:** Provide a concise description of the changes made in this pull request.
- [x] **Changelog:** Ensure a changelog entry following the format of [Keep a Changelog](https://keepachangelog.com/) is added at the bottom of the PR description.
- [x] **Documentation:** Have you updated relevant documentation [Open WebUI Docs](https://github.com/open-webui/docs), or other documentation sources?
- [x] **Dependencies:** Are there any new dependencies? Have you updated the dependency versions in the documentation?
- [x] **Testing:** Have you written and run sufficient tests to validate the changes?
- [x] **Code review:** Have you performed a self-review of your code, addressing any coding standard issues and ensuring adherence to the project's coding standards?
- [x] **Prefix:** To clearly categorize this pull request, prefix the pull request title using one of the following:
  - **BREAKING CHANGE**: Significant changes that may affect compatibility
  - **build**: Changes that affect the build system or external dependencies
  - **ci**: Changes to our continuous integration processes or workflows
  - **chore**: Refactor, cleanup, or other non-functional code changes
  - **docs**: Documentation update or addition
  - **feat**: Introduces a new feature or enhancement to the codebase
  - **fix**: Bug fix or error correction
  - **i18n**: Internationalization or localization changes
  - **perf**: Performance improvement
  - **refactor**: Code restructuring for better maintainability, readability, or scalability
  - **style**: Changes that do not affect the meaning of the code (white space, formatting, missing semi-colons, etc.)
  - **test**: Adding missing tests or correcting existing tests
  - **WIP**: Work in progress, a temporary label for incomplete or ongoing work

# Changelog Entry

### Description

- Optimized the `get_all_models()` function to reduce run time from 1.5~ seconds to 0.54~ seconds (in my environment) for about a 60%~ speed improvement. 
- The 0.5~ seconds is mostly from the `get_all_base_models` method.

### Added

- `Functions.get_functions_by_ids()` to batch multiple function lookups by id
- Dictionary-based lookups for custom models and functions
- Error handling for function module loading

### Changed

- Replaced individual database calls with batch queries using `Functions.get_functions_by_ids()`
- Converted lists operations to sets for faster speed.
- Pre-calculate action/filter IDs and store temporarily to avoid redundant processing
- Restructured custom model processing logic to use dictionary lookups instead of nested loops
- Modified function module loading to use centralized caching mechanism
- Replaced list concatenation with `extend()` method for better memory efficiency

### Deprecated

- None

### Removed

- Redundant database queries within model processing loops
- Repeated function module loading operations
- Inefficient list management / concatenation

### Breaking Changes

- **NONE I FOUND** - All changes are just re-writes and optimizations of the original code, in theory the input -> output should be exactly the same. The function input parameters, and output structure remain unchanged.

---

### Additional Information

- ~~**This code likely works though I haven't thoroughly tested it.**~~
- Tested the code locally, works great and its consistently faster. Editing models, filters, functions and pipes work.

### Contributor License Agreement

By submitting this pull request, I confirm that I have read and fully agree to the [Contributor License Agreement (CLA)](/CONTRIBUTOR_LICENSE_AGREEMENT), and I am providing my contributions under its terms.

torisetxd committed:
Add function to retrieve functions by IDs and optimize model fetching logic

RobertoGongora commented:
Why are you removing comments in unrelated parts of the code, and changing how certain variables are written with no effect? Feels done by AI.

Also the CI is failing so first get that working before asking for a review

torisetxd commented:
> Why are you removing comments in unrelated parts of the code, and changing how certain variables are written with no effect? Feels done by AI.
> 
> Also the CI is failing so first get that working before asking for a review
 
I removed the comments by accident since i'm used to removing comments in code since I personally don't like them. I didn't realize I removed them. I can re-open the PR or just make a new commit with the comments back and only the updated method.
 
This pr can also be closed I don't really mind. I was just offering an enhancement which greatly improved speed and just simply works. Sorry about the comments.

Also for the CI so many pr's on this repo have this fail and still get merged.
